Add CPO DRex computation

MFraters opened the pull request:
Add initial D-Rex code as close as possible to the ASPECT CPO/ D-Rex paper, but with a different random number generator. This also already include the fabric selector based on Karato 2008.

I had to think a bit about the structure to both keep the main code and unit test code reasonable. I also thought about generalizing the current 4 slip planes to n slip planes (it is an actual use case), but I think it is better to leave that for now as it should not be hard to do later.

I also added a CPO cite tag to the initialization and the references to the ASPECT CPO/D-Rex paper. Since D-Rex is a crucial part of it, I am considering also adding the 2 D-Rex papers to the list of requested citations. Is there any policy on that?

(point 2 in #3885)

tjhei commented on doc/make_cite_html.py:
```diff
@@ -72,6 +72,7 @@
                "fraters_menno_2020_3900603" : "GWB",
                "Fraters2019c" : "GWB",
                "FBTGS19" : "NewtonSolver",
+               "fraters_billen_2021_cpo", "CPO"
```
```suggestion
               "fraters_billen_2021_cpo": "CPO"
```

tjhei commented on doc/make_cite_html.py:
```diff
@@ -103,6 +104,7 @@
         "gassmoller2020formulations" : "Rene Gassmöller, Juliane Dannberg, Wolfgang Bangerth, Timo Heister, and Robert Myhill. 2020. “On Formulations of Compressible Mantle Convection.” Geophysical Journal International 221 (2) (February 13): 1264–1280. doi:10.1093/gji/ggaa078. http://dx.doi.org/10.1093/gji/ggaa078.",
         "heister_aspect_methods2" : "Timo Heister, Juliane Dannberg, Rene Gassmöller, and Wolfgang Bangerth. 2017. “High Accuracy Mantle Convection Simulation through Modern Numerical Methods – II: Realistic Models and Problems.” Geophysical Journal International 210 (2) (May 9): 833–851. doi:10.1093/gji/ggx195. http://dx.doi.org/10.1093/gji/ggx195.",
         "rose_freesurface" : "Ian Rose, Bruce Buffett, and Timo Heister. 2017. “Stability and Accuracy of Free Surface Time Integration in Viscous Flows.” Physics of the Earth and Planetary Interiors 262 (January): 90–100. doi:10.1016/j.pepi.2016.11.007. http://dx.doi.org/10.1016/j.pepi.2016.11.007."
+        "fraters_billen_2021_cpo" : "Fraters, M. R. T. and Billen, M. I. 2021. “On the Implementation and Usability of Crystal Preferred Orientation Evolution in Geodynamic Modeling” Geochemistry, Geophysics, Geosystems 22 (10): e2021GC009846. doi:10.1029/2021GC009846. https://doi.org/10.1029/2021GC009846."
```
missing comma before

tjhei commented on doc/sphinx/references.bib:
```diff
@@ -11590,6 +11590,22 @@ @Article{fraters:etal:2019
     eprint = {http://oup.prod.sis.lan/gji/article-pdf/218/2/873/28693654/ggz183.pdf},
 }
 
+@article{fraters:billen:2021,
```
The keys in the dictionary in the make_cite_html.py file do not match this bibtex key. Use either ":" or "_" consistently.

tjhei commented on include/aspect/particle/property/crystal_preferred_orientation.h:
```diff
@@ -37,12 +37,12 @@ namespace aspect
       enum class DeformationType
       {
-        passive
+        passive, olivine_a_fabric, olivine_b_fabric, olivine_c_fabric, olivine_d_fabric, olivine_e_fabric, enstatite
```
I know that we don't have documentation in the other places here either, but it would be nice to have short doxygen documentation what each of the enums means.

tjhei commented on include/aspect/particle/property/crystal_preferred_orientation.h:
```diff
@@ -222,6 +240,39 @@ namespace aspect
+          /**
+           * @brief Deterimens the deformation type from the deformation type selector.
+           *
+           * @param stress
```
This documentation is incomplete and empty. Either complete it or remove it, otherwise doxygen will warn.

tjhei commented on include/aspect/particle/property/crystal_preferred_orientation.h:
```diff
@@ -222,6 +240,39 @@ namespace aspect
+           * @brief Computes the deformation type given the stress and water content according to the
+           * table in Karato 2008.
+           *
+           * @param stress
```
same here

tjhei commented on include/aspect/particle/property/crystal_preferred_orientation.h:
```diff
@@ -222,6 +240,39 @@ namespace aspect
+                                                 const double water_content) const;
+
+          std::array<double,4>
+          reference_resolved_shear_stress_from_deformation_type(DeformationType deformation_type,
```
maybe document ``maxvalue``? units?

MFraters replied:
It returns the relative strength of the slip planes, which for olivine are 4. This is (more or less) how it is done in the fortran code. I have been thinking to generalize it (some minerals have way more slip planes), but I would rather to do that in a follow up pull request, because this is probably already complicated enough as it is. I will add better documentation.

tjhei commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -554,6 +614,400 @@ namespace aspect
+                                                                      const bool prevent_nondimensionalization) const
+      {
+        SymmetricTensor<2,3> strain_rate_nondimensional = strain_rate_3d;
+        Tensor<2,3> velocity_gradient_tensor_nondimensional = velocity_gradient_tensor;
```
Can you move those two linear below the if statement and always do ``= strain_rate_3d / nondimensionalization_value;``?

tjhei commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -554,6 +614,400 @@ namespace aspect
+        std::vector<double> strain_energy(n_grains);
+        double mean_strain_energy = 0;
+
+        // loop over grains
```
remove this comment (doesn't give additional information)

tjhei commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -554,6 +614,400 @@ namespace aspect
+            // compute G and beta
+            Tensor<1,4> bigI;
+            // this should be equal to a_cosine_matrices[grain_i]*a_cosine_matrices[grain_i]?
+            // todo: check and maybe replace?
```
todo?

tjhei commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -554,6 +614,400 @@ namespace aspect
+              {
+                for (unsigned int j = 0; j < 3; ++j)
+                  {
+                    bigI[0] = bigI[0] + strain_rate_nondimensional[i][j] * rotation_matrix[0][i] * rotation_matrix[1][j];
```
Is there maybe a tensor operator that does this multiplication?

MFraters replied:
Yes, it is taking the outer product of the slip system normal and slip direction, and then taking the scalar product of that with the non dimensional strain-rate. I have replaced the code and the test still works :)

tjhei commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -554,6 +614,400 @@ namespace aspect
+
+                index_inactive_q = std::distance(q_abs.begin(),max_element(q_abs.begin(), q_abs.end()));
+
+                // todo: explain
```
yes please

tjhei commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -554,6 +614,400 @@ namespace aspect
+                Assert(bigI[index_max_q] != 0.0, ExcMessage("Internal error: bigI is zero."));
+                double ratio = tau[index_max_q]/bigI[index_max_q];
+
+                double q_intermediate = ratio * (bigI[index_intermediate_q]/tau[index_intermediate_q]);
```
add const where ever you can.

tjhei commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -554,6 +614,400 @@ namespace aspect
+              {
+                // Following the Drex code, which differs from EPSL paper,
+                // which says gamma_nu depends on i+1: actually uses i+2
+                unsigned int ip2 = i + 2;
+                if (ip2 > 2)
+                  ip2 = ip2-3;
```
``const unsigned int ip2 = (i<=2) ? (i+2) : (i-1);`` ?

tjhei commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -554,6 +614,400 @@ namespace aspect
+                  }
+              }
+            // see comment on if all BigI are zero. In that case gamma should be zero.
+            double gamma = bottom != 0.0 ? top/bottom : 0;
```
```suggestion
            const double gamma = (bottom != 0.0) ? (top/bottom) : 0.0;
```

tjhei commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -554,6 +614,400 @@ namespace aspect
+
+            // compute w (equation 8, Kaminiski & Ribe, 2001)
+            // todo: explain what w is
+            // todo: there was a loop around this in the phyton code, discuss/check
```
todo

tjhei commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -554,6 +614,400 @@ namespace aspect
+              break;
+
+            default:
+              break;
```
right default?

tjhei commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -642,6 +1096,47 @@ namespace aspect
+                                   "Karato 2008 selector selects a fabric based on stress and water content as defined in "
+                                   "figure 4 of the Karato 2008 review paper (doi: 10.1146/annurev.earth.36.031207.124120).");
+
+                prm.declare_entry ("Mobility", "50",
```
can you add units to all parameters?

tjhei commented on unit_tests/crystal_preferred_orientation.cc:
```diff
@@ -511,3 +525,563 @@ TEST_CASE("CPO core: Spin tensor")
 }
+
+
+TEST_CASE("Fabric determination function")
```
very nice use of the unit tests!

MFraters commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -263,24 +288,27 @@ namespace aspect
-                                                                                            velocity_gradient_3d,
-                                                                                            ref_resolved_shear_stress);
+            std::pair<std::vector<double>, std::vector<Tensor<2,3>>>
+            derivatives_grains = this->compute_derivatives(data_position,
```
Yes, just below in the advection method block :)

MFraters commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -532,6 +567,31 @@ namespace aspect
+                                                                            compressible_strain_rate,
+                                                                            water_content);
+
+              set_deformation_type(cpo_index,data,mineral_i,static_cast<unsigned int>(DeformationType::passive));
```
ah, that was a mistake, thanks for catching it!

MFraters commented:
Hey @tjhei, Thanks for the review! I think I have addressed all your comments. Based on some of your comments I realized that it would probably be better to do some more groundwork for the generalization from 4 slip systems to n slip systems.

tjhei commented on include/aspect/particle/property/crystal_preferred_orientation.h:
```diff
@@ -35,21 +35,56 @@ namespace aspect
+       * enstatite: Only to be used with the D-Rex CPO Derivative algorithm. Sets the deformation type of the mineral to a enstatite Fabric, which influences the relative strenght of the slip planes.
+       * olivine_karato_2008: Only to be used with the D-Rex CPO Derivative algorithm. Sets the deformation type of the mineral to a olivine fabric based on the table in Karato 2008.
+       */
       enum class DeformationTypeSelector
```
What is the difference between DeformationType and DeformationTypeSelector?

MFraters commented:
@tjhei I have addressed your comments. I removed some of the repetition in the description of the enum and added some text to explain the difference between the deformation type and the deformation type selector. Can you take a look to see whether the text is more clear now?

gassmoeller commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -554,6 +614,369 @@ namespace aspect
+              ref_resolved_shear_stress[3] = 1;
+              break;
+
+            // from Kaminski and Ribe, GRL 2002and
```
```suggestion
            // from Kaminski and Ribe, GRL 2002 and
```

gassmoeller commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -554,6 +614,369 @@ namespace aspect
+              ref_resolved_shear_stress[3] = max_value;
+              break;
+
+            // Kaminski, Ribe and Browaeys, JGI, 2004 (same as in the matlab code)and
```
```suggestion
            // Kaminski, Ribe and Browaeys, GJI, 2004 (same as in the matlab code) and
```

gassmoeller commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -554,6 +614,369 @@ namespace aspect
+              break;
+
+            // from Kaminski and Ribe, GJI 2004.
+            // Todo: this one is not used in practice, since there is an optimalisation in
```
```suggestion
            // Todo: this one is not used in practice, since there is an optimization in
```

gassmoeller commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -642,6 +1065,45 @@ namespace aspect
+                                   Patterns::Double(0),
+                                   "The dimensionless intrinsic grain boundary mobility for both olivine and enstatite.");
+
+                prm.declare_entry ("Volume fractions minerals", "0.5, 0.5",
```
Is 0.5,0.5 a good default value? What is the proportion in the mantle?

MFraters replied:
You are right, given the other defaults, it have now set it to 0.7, 0.3.

gassmoeller commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -642,6 +1065,45 @@ namespace aspect
+                                   Patterns::Double(0),
+                                   "This is exponent p as defined in equation 11 of Kaminski et al., 2004. ");
+
+                prm.declare_entry ("Nucleation efficientcy", "5",
```
```suggestion
                prm.declare_entry ("Nucleation efficiency", "5",
```

gassmoeller commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -642,6 +1065,45 @@ namespace aspect
+
+                prm.declare_entry ("Threshold GBS", "0.3",
+                                   Patterns::Double(0),
+                                   "This is the dimensionless grain-boundary sliding threshold. ");
```
Can you elaborate what the threshold is for? I.e. is it a stress threshold? slip rate threshold?

gassmoeller commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -744,6 +1234,17 @@ namespace aspect
+                volume_fractions_minerals = Utilities::string_to_double(dealii::Utilities::split_string_list(prm.get("Volume fractions minerals")));
+                stress_exponent = prm.get_double("Stress exponents");
+                exponent_p = prm.get_double("Exponents p");
+                nucleation_efficientcy = prm.get_double("Nucleation efficientcy");
```
```suggestion
                nucleation_efficiency = prm.get_double("Nucleation efficiency");
```

gassmoeller reviewed (commented):
A bunch of comments :-). But mostly requests for more documentation, or suggestions to improve efficiency. Can you please at least add one test model that uses the new functionality. In particular there were typos in input parameters that were not caught, because no test actually uses the whole D-REX calculation together (the existing tests use the passive case without fabric evolution).

MFraters commented:
> Can you please at least add one test model that uses the new functionality. In particular there were typos in input parameters that were not caught, because no test actually uses the whole D-REX calculation together (the existing tests use the passive case without fabric evolution).

This is a bit of a problem of cutting up a pull request this much. I have tests for this which depend on other parts (need at least the postprocessor). That is why I put it as a seperate point (point 9) in the tracking issue #3885. What do you think is the best way to move forward on this?

And about the typos in the input parameters, don't worry, they are also definitely in the input files of the tests ;)

MFraters commented on include/aspect/particle/property/crystal_preferred_orientation.h:
```diff
@@ -35,21 +35,55 @@ namespace aspect
+       * @brief The type of deformation selector used by the CPO code.
+       *
+       * The selector is a input parameter and it can either set a deformation type directly or deterimine the deformation type through an algorithm.
+       * The deformation type selector is used to determine/select the the deformation type. It can be a fixed deformation type, for example,
```
For example for the `DeformationTypeSelector` `olivine_karato_2008`, the actual `DeformationType` can be `olivine_a_fabric`, or `olivine_a_fabric`, etc.

MFraters commented:
> (the existing tests use the passive case without fabric evolution).

There is a unit test which tests 2 different computations of the derivatives with an actual fabric (one of which I computed a analytic solution of the derivatives by hand). The adding computation of these derivatives is the core part of this pull request, since it is what the D-Rex algorithm essentially is.

MFraters commented on include/aspect/particle/property/crystal_preferred_orientation.h:
```diff
@@ -222,6 +274,38 @@ namespace aspect
           get_number_of_minerals() const;
 
+          /**
+           * @brief Deterimens the deformation type from the deformation type selector.
```
see my comment above on the difference

MFraters commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -227,6 +229,29 @@ namespace aspect
+        const double water_content = solution[this->introspection().component_indices.compositional_fields[water_idx]];
+
+        // get the composition of the particle
+        std::vector<double> compositions;
```
The reason I did that was because it would then need to be extracted within the loop over the minerals. Although my case are currently are only 2 minerals, but it might become inefficient for cases with many minerals. It would simplify the interface considerably though...

gassmoeller replied:
Well, the only thing that is faster in the current implementation than directly accessing the data from solution is that `this->introspection` is called less often. In other words the following two algorithms are really nearly identical in speed (but the second is much easier to read if you have a lot of indices and values like you do):
```
Vector<double> solution (some content);
const double pressure = solution[this->introspection().component_indices.pressure];

for (unsigned int i=0; i<50; ++i)
  sum += pressure;
```

and:

```
Vector<double> solution (some content);
const auto &introspection = this->introspection();

for (unsigned int i=0; i<50; ++i)
  sum += solution[introspection.component_indices.pressure];
```

In fact after compiler optimizations they may be the same anyway.

Anyway, I guess this is a long winded way of saying: If you like your implementation because you find it easier to read that is ok, we can keep it. But if you like it, because you guess that it is faster, I would say it probably is not. Also if you are concerned about speed: If I would need to guess the slowest part of your algorithm, I would suspect line 720 and 774, where you call std::pow for every mineral and grain and slip system. But even with this guess, I would not start any optimizations before I had measured its impact.

So bring this into whichever form you think is best to read for now, and we can worry about optimizations afterwards.

MFraters replied:
I can see readability advantages for both approaches, so I don't really have a strong preference necessarily. 

>  If I would need to guess the slowest part of your algorithm, I would suspect line 720 and 774, where you call std::pow for every mineral and grain and slip system.

haha, yes, everytime I profile stuff I keep being supersized how slow std::pow can be :)

I gave it a quick try to see if I would like it more the other way, and another problem quickly showed up. It introduces introspection calls into the function, which make it much harder to unit test the function. I am sure we could find a solution to make the unit tests work with these introspection calls, but I kinda like that the function is fully self contained. Let me know what you think.

MFraters commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -554,6 +614,369 @@ namespace aspect
+            // compute w (equation 8, Kaminiski & Ribe, 2001)
+            // w is the Rotation rate vector of the crystallographic axes of grain
+            w[0] = 0.5*(velocity_gradient_tensor_nondimensional[2][1]-velocity_gradient_tensor_nondimensional[1][2]) - 0.5*(G[2][1]-G[1][2])*gamma;
+            w[1] = 0.5*(velocity_gradient_tensor_nondimensional[0][2]-velocity_gradient_tensor_nondimensional[2][0]) - 0.5*(G[0][2]-G[2][0])*gamma;
```
The equation on which this is based (equation 8, Kaminiski & Ribe, 2001) states: 

$$
\begin{align}
\omega^v_i = \frac{L_{i+2 i+1}-L_{i+1 i+2}}{2} -  \frac{G^v_{i+2 i+1}-G^v_{i+1 i+2}}{2} \gamma_v
\end{align}
$$

where i = 1,2,3, but with c++ i = 0,1,2

With the symmetry I can see where you are coming from, and I had to look into it to understand again why it is done this way. The goal here is to get a rotation rate tensor $\bf \omega$. We can get a rotation rate tensor with the following: $W = \frac{L-L^T}{2}$. The next step is to convert this second order tensor to a vector. Looking at https://www.continuummechanics.org/velocitygradient.html, the rotation rate tensor $W$ can be summarized as the following: 

$$
W = \pmatrix{0 &-\omega_z &\omega_y \\
\omega_z &0 &-\omega_x \\
-\omega_y &\omega_x &0} = 
\left[ \matrix{
0 &
{1 \over 2} \left( {\partial v_x \over \partial y} - {\partial v_y \over \partial x} \right) & 
{1 \over 2} \left( {\partial v_x \over \partial z} - {\partial v_z \over \partial x} \right) \\
\\
{1 \over 2} \left( {\partial v_y \over \partial x} - {\partial v_x \over \partial y} \right) & 
0 &
{1 \over 2} \left( {\partial v_y \over \partial z} - {\partial v_z \over \partial y} \right) \\
\\
{1 \over 2} \left( {\partial v_z \over \partial x} - {\partial v_x \over \partial z} \right) & 
{1 \over 2} \left( {\partial v_z \over \partial y} - {\partial v_y \over \partial z} \right) &
0 
} \right]
$$

The equation for the elements of the vector is 

$$
\omega_x = {1 \over 2} \left( {\partial v_z \over \partial y} - {\partial v_y \over \partial z} \right) 
$$

$$
\omega_y  = {1 \over 2} \left( {\partial v_x \over \partial z} - {\partial v_z \over \partial x} \right)
$$

$$
\omega_z = {1 \over 2} \left( {\partial v_y \over \partial x} - {\partial v_x \over \partial y} \right)
$$

So due to the "asymmetry" in $W$ in terms of defining $\bf \omega$, you get an asymmetry in the construction of vector $\bf \omega$

gassmoeller replied:
Ok, thanks for checking!

MFraters commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -554,6 +614,369 @@ namespace aspect
+              {
+                // Following the Drex code, which differs from EPSL paper,
+                // which says gamma_nu depends on i+1: actually uses i+2
+                const unsigned int ip2 = (i==0) ? (i+2) : (i-1);
```
it does indeed stand for "i plus 2", but since it wraps around from i=1 -> ip2 = 0 and i=2 -> ip2 = 1. I guess the best name would be `i_plus_2_wrapped_around`, but that is a bit long. I have changed it to `i_offset`. It is a bit generic, but maybe more clear. What do you think?

MFraters commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -554,6 +614,369 @@ namespace aspect
+
+            // Now calculate the analytic solution to the deformation minimization problem
+            // compute gamma (equation 7, Kaminiski & Ribe, 2001)
+            double top = 0;
+            double bottom = 0;
```
They are the numerator and the denominator in equation equation 7 in Kaminiski & Ribe (2001), but I can never tell apart which is which by the names, so I just named them top and bottom. I will document it a bit better.

MFraters commented:
@gassmoeller Thanks for the review! I have addressed all your comments. I have left the comment I thought would be good for you to read if my reply is satisfactory unresolved.

gassmoeller commented on include/aspect/particle/property/crystal_preferred_orientation.h:
```diff
@@ -222,6 +274,38 @@ namespace aspect
           get_number_of_minerals() const;
 
+          /**
+           * @brief Determines the deformation type from the deformation type selector.
```
Would this correctly explain this function? If so, please include. If not, please include a correct explanation ;-).

```suggestion
           * @brief Determines the deformation type from the deformation type selector.
           * If the provided @p deformation_type_selector is a specific deformation type,
           * the function will simply return the corresponding deformation type. However,
           * if the @p deformation_type_selector is an algorithm to determine the current
           * deformation type (e.g. based on measured lab data or analytical models), then
           * the function computes the appropriate deformation type at the given conditions
           * and returns the compute deformation type.
```

gassmoeller reviewed (commented):
Thank you, and sorry for the delay, looks much better now. A few comments left.

MFraters reviewed (commented):
Thanks for the review. I addressed your comments. The only outstanding issue is about using `solution` and `gradients` in `compute_derivates`. Let me know what you think about that.

gassmoeller reviewed (approved):
> I gave it a quick try to see if I would like it more the other way, and another problem quickly showed up. It introduces introspection calls into the function, which make it much harder to unit test the function.

Ok, that is a good point. Let's not worry about it then for now.
